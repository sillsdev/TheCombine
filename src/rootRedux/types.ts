import { type Action } from "redux";
import { type ThunkDispatch } from "redux-thunk";

import {
  type LoginState,
  defaultState as loginState,
} from "components/Login/Redux/LoginReduxTypes";
import {
  type CurrentProjectState,
  defaultState as currentProjectState,
} from "components/Project/ProjectReduxTypes";
import {
  type ExportProjectState,
  defaultState as exportProjectState,
} from "components/ProjectExport/Redux/ExportProjectReduxTypes";
import {
  type PronunciationsState,
  defaultState as pronunciationsState,
} from "components/Pronunciations/Redux/PronunciationsReduxTypes";
import {
  type TreeViewState,
  defaultState as treeViewState,
} from "components/TreeView/Redux/TreeViewReduxTypes";
import {
  type CharacterInventoryState,
  defaultState as characterInventoryState,
} from "goals/CharacterInventory/Redux/CharacterInventoryReduxTypes";
import {
  type MergeTreeState,
  defaultState as mergeDuplicateGoal,
} from "goals/MergeDuplicates/Redux/MergeDupsReduxTypes";
import {
  type GoalsState,
  defaultState as goalsState,
} from "goals/Redux/GoalReduxTypes";
import {
  type AnalyticsState,
  defaultState as analyticsState,
} from "types/Redux/analyticsReduxTypes";

/** Redux store structure */
export interface StoreState {
  //login
  loginState: LoginState;

  //project
  currentProjectState: CurrentProjectState;
  exportProjectState: ExportProjectState;

  //data entry and review entries
  treeViewState: TreeViewState;
  pronunciationsState: PronunciationsState;

  //goal timeline and current goal
  goalsState: GoalsState;

  //merge duplicates goal
  mergeDuplicateGoal: MergeTreeState;

  //character inventory goal
  characterInventoryState: CharacterInventoryState;

  //analytics state
  analyticsState: AnalyticsState;
}

/** Default values for the Redux store */
export const defaultState: StoreState = {
  loginState: { ...loginState },
  currentProjectState: { ...currentProjectState },
  exportProjectState: { ...exportProjectState },
  treeViewState: { ...treeViewState },
  pronunciationsState: { ...pronunciationsState },
  goalsState: { ...goalsState },
  mergeDuplicateGoal: { ...mergeDuplicateGoal },
  characterInventoryState: { ...characterInventoryState },
  analyticsState: { ...analyticsState },
};

export interface ActionWithPayload<T> extends Action {
  payload: T;
}

// https://redux.js.org/usage/usage-with-typescript#type-checking-redux-thunks
// suggests a custom general type for ThunkAction,
// so in like fashion, here's one for ThunkDispatch:
export type StoreStateDispatch = ThunkDispatch<StoreState, any, Action>;
