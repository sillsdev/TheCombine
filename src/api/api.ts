/* tslint:disable */
/* eslint-disable */
/**
 * BackendFramework
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

export * from "./api/audio-api";
export * from "./api/avatar-api";
export * from "./api/invite-api";
export * from "./api/lift-api";
export * from "./api/merge-api";
export * from "./api/project-api";
export * from "./api/user-api";
export * from "./api/user-edit-api";
export * from "./api/user-role-api";
export * from "./api/word-api";
