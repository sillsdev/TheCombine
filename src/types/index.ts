import { LoginState } from "components/Login/Redux/LoginReduxTypes";
import { PasswordResetState } from "components/PasswordReset/Redux/ResetReduxTypes";
import { CurrentProjectState } from "components/Project/ProjectReduxTypes";
import { ExportProjectState } from "components/ProjectExport/Redux/ExportProjectReduxTypes";
import { CreateProjectState } from "components/ProjectScreen/CreateProject/Redux/CreateProjectReduxTypes";
import { PronunciationsState } from "components/Pronunciations/Redux/PronunciationsReduxTypes";
import { TreeViewState } from "components/TreeView/TreeViewReducer";
import { CharacterInventoryState } from "goals/CharInventoryCreation/Redux/CharacterInventoryReduxTypes";
import { MergeTreeState } from "goals/MergeDupGoal/Redux/MergeDupReduxTypes";
import { ReviewEntriesState } from "goals/ReviewEntries/ReviewEntriesComponent/Redux/ReviewEntriesReduxTypes";
import { AnalyticsState } from "types/Redux/analyticsReduxTypes";
import { GoalsState } from "types/goals";

//root store structure
export interface StoreState {
  //login
  loginState: LoginState;
  passwordResetState: PasswordResetState;

  //project
  createProjectState: CreateProjectState;
  currentProjectState: CurrentProjectState;
  exportProjectState: ExportProjectState;

  //data entry and review entries
  treeViewState: TreeViewState;
  reviewEntriesState: ReviewEntriesState;
  pronunciationsState: PronunciationsState;

  //goal timeline and current goal
  goalsState: GoalsState;

  //merge duplicates goal
  mergeDuplicateGoal: MergeTreeState;

  //character inventory goal
  characterInventoryState: CharacterInventoryState;

  //analytics state
  analyticsState: AnalyticsState;
}
