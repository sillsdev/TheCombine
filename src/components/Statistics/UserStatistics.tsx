import { Card, Grid, ListItem, List } from "@mui/material";
import { ReactElement, useState, useEffect } from "react";

import { SemanticDomainUserCount } from "api/models";
import { getSemanticDomainUserCount } from "backend";
import * as LocalStorage from "backend/localStorage";
import { ColumnHead, TableCell } from "components/Statistics/TableCells";

interface UserStatisticsProps {
  lang: string;
}

export default function UserStatistics(
  props: UserStatisticsProps
): ReactElement {
  const [domainUserCountList, setDomainUserCountList] = useState<
    SemanticDomainUserCount[]
  >([]);

  useEffect(() => {
    const updateSemanticDomainUserCounts = async (): Promise<void> => {
      const counts = await getUserStatistics(
        LocalStorage.getProjectId(),
        props.lang
      );
      if (counts !== undefined) {
        return setDomainUserCountList(counts);
      }
    };
    updateSemanticDomainUserCounts();
  }, [props.lang]);

  async function getUserStatistics(
    projectId: string,
    lang?: string
  ): Promise<SemanticDomainUserCount[] | undefined> {
    return await getSemanticDomainUserCount(projectId, lang);
  }

  return (
    <Grid container justifyContent="center">
      <Card style={{ width: 600 }}>
        <List>
          <Grid container wrap="nowrap" justifyContent="space-around">
            <ColumnHead titleId={"statistics.column.username"} />
            <ColumnHead titleId={"statistics.column.domainCount"} />
            <ColumnHead titleId={"statistics.column.senseCount"} />
          </Grid>
        </List>
        <List>
          {domainUserCountList.map((t) => (
            <TableRow key={t.id} counts={t} />
          ))}
        </List>
      </Card>
    </Grid>
  );
}

function TableRow(props: { counts: SemanticDomainUserCount }): ReactElement {
  return (
    <ListItem style={{ minWidth: "600px" }}>
      <Grid container wrap="nowrap" justifyContent="space-around">
        <TableCell text={props.counts.username} />
        <TableCell text={props.counts.domainCount} />
        <TableCell text={props.counts.wordCount} />
      </Grid>
    </ListItem>
  );
}
