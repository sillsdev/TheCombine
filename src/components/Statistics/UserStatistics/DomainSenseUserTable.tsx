import { Grid, Typography } from "@material-ui/core";
import React, { ReactElement } from "react";

import { DomainSenseUserCount } from "api/models";

interface StatisticsProps {
  domainSenseUserCount: DomainSenseUserCount;
}

export default function StatisticsTable(props: StatisticsProps): ReactElement {
  return (
    <React.Fragment>
      <Grid container wrap="nowrap" justifyContent="space-around">
        <Grid
          item
          xs={5}
          key={"id_" + props.domainSenseUserCount.id}
          style={{
            borderBottomStyle: "dotted",
            borderBottomWidth: 1,
            position: "relative",
          }}
        >
          <Typography variant="body1">
            {props.domainSenseUserCount.username}
          </Typography>
        </Grid>
        <Grid
          item
          xs={5}
          key={"domain_" + props.domainSenseUserCount.id}
          style={{
            borderBottomStyle: "dotted",
            borderBottomWidth: 1,
            position: "relative",
          }}
        >
          <Typography variant="body1">
            {props.domainSenseUserCount.domainCount}
          </Typography>
        </Grid>
        <Grid
          item
          xs={5}
          key={"statistics_" + props.domainSenseUserCount.id}
          style={{
            borderBottomStyle: "dotted",
            borderBottomWidth: 1,
            position: "relative",
          }}
        >
          <Typography variant="body1">
            {props.domainSenseUserCount.senseCount}
          </Typography>
        </Grid>
      </Grid>
    </React.Fragment>
  );
}
