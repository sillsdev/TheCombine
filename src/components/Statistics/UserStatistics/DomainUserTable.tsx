import { Grid, Typography } from "@material-ui/core";
import React, { ReactElement } from "react";

import { SemanticDomainUserCount } from "api/models";

interface StatisticsProps {
  semanticDomainUserCount: SemanticDomainUserCount;
}

export default function StatisticsTable(props: StatisticsProps): ReactElement {
  return (
    <React.Fragment>
      <Grid container wrap="nowrap" justifyContent="space-around">
        <Grid
          item
          xs={5}
          key={"id_" + props.semanticDomainUserCount.id}
          style={{
            borderBottomStyle: "dotted",
            borderBottomWidth: 1,
            position: "relative",
          }}
        >
          <Typography variant="body1">
            {props.semanticDomainUserCount.username}
          </Typography>
        </Grid>
        <Grid
          item
          xs={5}
          key={"domain_" + props.semanticDomainUserCount.id}
          style={{
            borderBottomStyle: "dotted",
            borderBottomWidth: 1,
            position: "relative",
          }}
        >
          <Typography variant="body1">
            {props.semanticDomainUserCount.domainCount}
          </Typography>
        </Grid>
        <Grid
          item
          xs={5}
          key={"statistics_" + props.semanticDomainUserCount.id}
          style={{
            borderBottomStyle: "dotted",
            borderBottomWidth: 1,
            position: "relative",
          }}
        >
          <Typography variant="body1">
            {props.semanticDomainUserCount.wordCount}
          </Typography>
        </Grid>
      </Grid>
    </React.Fragment>
  );
}
