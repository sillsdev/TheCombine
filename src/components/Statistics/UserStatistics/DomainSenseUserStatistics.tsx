import { Card, Grid, Typography, ListItem, List } from "@material-ui/core";
import React, { ReactElement, useState, useEffect } from "react";
import { useTranslation } from "react-i18next";

import DomainSenseUserTable from "./DomainSenseUserTable";
import { DomainSenseUserCount, Project } from "api/models";
import { getDomainSenseUserCounts } from "backend";
import * as LocalStorage from "backend/localStorage";

interface DomainSenseUserStatisticsProps {
  currentProject?: Project;
  lang: string;
}

export default function DomainSenseUserStatistics(
  props: DomainSenseUserStatisticsProps
): ReactElement {
  const [domainSenseUserList, setDomainSenseUser] = useState<
    DomainSenseUserCount[]
  >([]);
  const { t } = useTranslation();

  useEffect(() => {
    const updateDomainSenseUserCount = async () => {
      const counts = await getDomainSenseUserStatistics(
        LocalStorage.getProjectId(),
        props.lang
      );
      if (counts != undefined) {
        return setDomainSenseUser(counts);
      }
    };
    updateDomainSenseUserCount();
  }, [props.lang]);

  async function getDomainSenseUserStatistics(
    projectId: string,
    lang?: string
  ): Promise<DomainSenseUserCount[] | undefined> {
    return await getDomainSenseUserCounts(projectId, lang);
  }

  function getDomainSenseUserList(domainSenseUser: DomainSenseUserCount[]) {
    return domainSenseUser.map((t) => (
      <ListItem style={{ minWidth: "600px" }} key={`${t.id}`}>
        <DomainSenseUserTable key={`${t.id}`} domainSenseUserCount={t!} />
      </ListItem>
    ));
  }

  return (
    <React.Fragment>
      <Grid container justifyContent="center">
        <Card style={{ width: 600 }}>
          <List>
            <Grid container wrap="nowrap" justifyContent="space-around">
              <Grid
                item
                xs={5}
                style={{
                  borderBottomStyle: "dotted",
                  borderBottomWidth: 1,
                  position: "relative",
                }}
              >
                <Typography variant="subtitle1">
                  {t("statistics.userName")}
                </Typography>
              </Grid>
              <Grid
                item
                xs={5}
                style={{
                  borderBottomStyle: "dotted",
                  borderBottomWidth: 1,
                  position: "relative",
                }}
              >
                <Typography variant="subtitle1">
                  {t("statistics.domainCount")}
                </Typography>
              </Grid>
              <Grid
                item
                xs={5}
                style={{
                  borderBottomStyle: "dotted",
                  borderBottomWidth: 1,
                  position: "relative",
                }}
              >
                <Typography variant="subtitle1">
                  {t("statistics.senseCount")}
                </Typography>
              </Grid>
            </Grid>
          </List>
          <List>{getDomainSenseUserList(domainSenseUserList)}</List>
        </Card>
      </Grid>
    </React.Fragment>
  );
}
