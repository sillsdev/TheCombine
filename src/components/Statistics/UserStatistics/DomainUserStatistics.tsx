import { Card, Grid, Typography, ListItem, List } from "@mui/material";
import React, { ReactElement, useState, useEffect } from "react";
import { useTranslation } from "react-i18next";

import DomainSenseUserTable from "./DomainUserTable";
import { SemanticDomainUserCount, Project } from "api/models";
import { getSemanticDomainUserCount, getFrontierWords } from "backend";
import * as LocalStorage from "backend/localStorage";

interface DomainSenseUserStatisticsProps {
  currentProject?: Project;
  lang: string;
}

export default function DomainSenseUserStatistics(
  props: DomainSenseUserStatisticsProps
): ReactElement {
  const [domainUserCountList, setDomainUserCountList] = useState<
    SemanticDomainUserCount[]
  >([]);
  const { t } = useTranslation();

  useEffect(() => {
    const updateSemanticDomainUserCounts = async () => {
      const counts = await getUserStatistics(
        LocalStorage.getProjectId(),
        props.lang
      );
      if (counts != undefined) {
        return setDomainUserCountList(counts);
      }
    };
    updateSemanticDomainUserCounts();
  }, [props.lang]);

  async function getUserStatistics(
    projectId: string,
    lang?: string
  ): Promise<SemanticDomainUserCount[] | undefined> {
    return await getSemanticDomainUserCount(projectId, lang);
  }

  function getDomainSenseUserList(
    semanticDomainUserCount: SemanticDomainUserCount[]
  ) {
    return semanticDomainUserCount.map((t) => (
      <ListItem style={{ minWidth: "600px" }} key={`${t.id}`}>
        <DomainSenseUserTable key={`${t.id}`} semanticDomainUserCount={t!} />
      </ListItem>
    ));
  }

  return (
    <React.Fragment>
      <Grid container justifyContent="center">
        <Card style={{ width: 600 }}>
          <List>
            <Grid container wrap="nowrap" justifyContent="space-around">
              <Grid
                item
                xs={5}
                style={{
                  borderBottomStyle: "dotted",
                  borderBottomWidth: 1,
                  position: "relative",
                }}
              >
                <Typography variant="subtitle1">
                  {t("statistics.userName")}
                </Typography>
              </Grid>
              <Grid
                item
                xs={5}
                style={{
                  borderBottomStyle: "dotted",
                  borderBottomWidth: 1,
                  position: "relative",
                }}
              >
                <Typography variant="subtitle1">
                  {t("statistics.domainCount")}
                </Typography>
              </Grid>
              <Grid
                item
                xs={5}
                style={{
                  borderBottomStyle: "dotted",
                  borderBottomWidth: 1,
                  position: "relative",
                }}
              >
                <Typography variant="subtitle1">
                  {t("statistics.senseCount")}
                </Typography>
              </Grid>
            </Grid>
          </List>
          <List>{getDomainSenseUserList(domainUserCountList)}</List>
        </Card>
      </Grid>
    </React.Fragment>
  );
}
