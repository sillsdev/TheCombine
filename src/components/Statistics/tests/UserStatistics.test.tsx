import { ListItem } from "@mui/material";
import renderer from "react-test-renderer";

import "tests/reactI18nextMock";

import { SemanticDomainUserCount } from "api";
import UserStatistics from "components/Statistics/UserStatistics";
import { newSemanticDomainUserCount } from "types/semanticDomain";

let testRenderer: renderer.ReactTestRenderer;

const mockProjectId = "mockProjectId";
const mockSemanticDomainUserCount = newSemanticDomainUserCount();
const mockSemanticDomainUserCountArray: Array<SemanticDomainUserCount> = [
  mockSemanticDomainUserCount,
];

const mockGetDomainSenseUserStatistics = jest.fn();
const mockGetProjectId = jest.fn();

jest.mock("backend", () => ({
  getSemanticDomainUserCount: (projectId: string, lang?: string) =>
    mockGetDomainSenseUserStatistics(projectId, lang),
}));

jest.mock("backend/localStorage", () => ({
  getProjectId: () => mockGetProjectId(),
}));

function setMockFunctions() {
  mockGetProjectId.mockReturnValue(mockProjectId);
  mockGetDomainSenseUserStatistics.mockResolvedValue(
    mockSemanticDomainUserCountArray
  );
}

beforeEach(async () => {
  jest.clearAllMocks();
  setMockFunctions();
  await renderer.act(async () => {
    testRenderer = renderer.create(<UserStatistics lang={""} />);
  });
});

describe("UserStatistics", () => {
  it("renders without crashing, UI does not change unexpectedly", async () => {
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it("useEffect hook was called", async () => {
    //Verify the mock function called
    expect(mockGetProjectId).toBeCalled();

    //Verify ListItem for the DomainSenseUserCount object is present
    const newSenDomCountList = testRenderer.root.findAllByType(ListItem);
    expect(newSenDomCountList.length).toEqual(
      mockSemanticDomainUserCountArray.length
    );
  });
});
