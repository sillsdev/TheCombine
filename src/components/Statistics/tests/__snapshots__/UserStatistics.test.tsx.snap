// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserStatistics renders without crashing, UI does not change unexpectedly 1`] = `
<div
  className="MuiGrid-root MuiGrid-container css-1lym95h-MuiGrid-root"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    style={
      Object {
        "width": 600,
      }
    }
  >
    <ul
      className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-wrap-xs-nowrap css-1ymxtd0-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 css-k3f0g4-MuiGrid-root"
          style={
            Object {
              "borderBottomStyle": "dotted",
              "borderBottomWidth": 1,
              "position": "relative",
            }
          }
        >
          <h6
            className="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
          >
            statistics.column.username
          </h6>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 css-k3f0g4-MuiGrid-root"
          style={
            Object {
              "borderBottomStyle": "dotted",
              "borderBottomWidth": 1,
              "position": "relative",
            }
          }
        >
          <h6
            className="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
          >
            statistics.column.domainCount
          </h6>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 css-k3f0g4-MuiGrid-root"
          style={
            Object {
              "borderBottomStyle": "dotted",
              "borderBottomWidth": 1,
              "position": "relative",
            }
          }
        >
          <h6
            className="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
          >
            statistics.column.senseCount
          </h6>
        </div>
      </div>
    </ul>
    <ul
      className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
        disabled={false}
        style={
          Object {
            "minWidth": "600px",
          }
        }
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-wrap-xs-nowrap css-1ymxtd0-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 css-k3f0g4-MuiGrid-root"
            style={
              Object {
                "borderBottomStyle": "dotted",
                "borderBottomWidth": 1,
                "position": "relative",
              }
            }
          >
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            />
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 css-k3f0g4-MuiGrid-root"
            style={
              Object {
                "borderBottomStyle": "dotted",
                "borderBottomWidth": 1,
                "position": "relative",
              }
            }
          >
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            />
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 css-k3f0g4-MuiGrid-root"
            style={
              Object {
                "borderBottomStyle": "dotted",
                "borderBottomWidth": 1,
                "position": "relative",
              }
            }
          >
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            />
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
